src/FrameDecorator/Table.php

```diff
@@ -232,143 +232,190 @@ public function set_max_width($width)
+        foreach ($children as $child) {
+            $display = $child->get_style()->display;
 
-            } else {
+            if (in_array($display, self::$ROW_GROUPS, true)) {
+                // Reset anonymous tbody
+                $tbody = null;
 
-                if ($display === "table-row") {
-                    $child->normalise();
-                    continue;
+                // Add headers and footers
+                if ($display === "table-header-group") {
+                    $this->_headers[] = $child;
+                } elseif ($display === "table-footer-group") {
+                    $this->_footers[] = $child;
                 }
+                continue;
+            }
 
-                if ($display === "table-cell") {
-                    $css = $this->get_style()->get_stylesheet();
+            if (in_array($display, $column_caption, true)) {
+                continue;
+            }
 
-                    // Create an anonymous table row group
-                    $tbody = $this->get_node()->ownerDocument->createElement("tbody");
+            // Remove empty text nodes between valid children
+            if ($this->isEmptyTextNode($child)) {
+                $this->remove_child($child);
+                continue;
+            }
 
-                    $frame = new Frame($tbody);
+            // Catch consecutive misplaced frames within a single anonymous group
+            if ($tbody === null) {
+                $tbody = $this->createFrame("tbody", "table-row-group");
+                $this->insert_child_before($tbody, $child);
+            }
 
-                    $style = $css->create_style();
-                    $style->inherit($this->get_style());
+            $tbody->append_child($child);
+        }
 
-                    // Lookup styles for tbody tags.  If the user wants styles to work
-                    // better, they should make the tbody explicit... I'm not going to
-                    // try to guess what they intended.
-                    foreach ($css->lookup("tbody") as $tbody_style) {
-                        $style->merge($tbody_style);
-                    }
-                    $style->display = "table-row-group";
+        foreach ($this->get_children() as $child) {
+            $display = $child->get_style()->display;
 
-                    // Okay, I have absolutely no idea why I need this clone here, but
-                    // if it's omitted, php (as of 2004-07-28) segfaults.
-                    $frame->set_style($style);
-                    $table_row_group = Factory::decorate_frame($frame, $this->_dompdf, $this->_root);
+            if (in_array($display, self::$ROW_GROUPS, true)) {
+                $this->normalize_group($child);
+            }
+        }
+    }
 
-                    // Create an anonymous table row
-                    $tr = $this->get_node()->ownerDocument->createElement("tr");
+    private function normalize_group(AbstractFrameDecorator $frame): void
```

Comment:
Should the child normalization methods go into the relevant FrameDecorator class (TableRowGroup, TableRow)?